`OSSL_CMP_CTX_reinit()`: fix missing reset of `ctx->genm_ITAVs` and of `ctx->geninfo_ITAVs`

DDvO opened the pull request:
Otherwise, further `OSSL_CMP_exec_GENM_ses()` calls go wrong.

slontis reviewed (commented):
Please add test

DDvO commented:
> Please add test

Done.
I made the CMP mock server more realistic by checking at least the number of ITAVS in a GENM request.
Without the fix, the extended `execute_exec_GENM_ses_test()`, which now does two GENM sessions, fails.

DDvO commented:
On this occasion I found and fixed an outdated example in `OSSL_CMP_CTX_new.pod`.

DDvO commented:
This time there are some strange unrelated CI failures on `01-test_symbol_presence.t`.

slontis commented:
The failures are happening on all builds, so not your problem.

DDvO commented:
Thanks @t8m for your swift approval.

I just realized that an update of the doc of `OSSL_CMP_CTX_reinit()` was missing,
the new check in `cmp_mock_srv.c` should be relaxed a little,

DDvO commented:
It turns out that a similar issue was with `ctx->geninfo_ITAVs`,
where in this case the solution is to add a missing API function `OSSL_CMP_CTX_reset_geninfo_ITAVs()` because the `generalInfo` header entries (such as `certProfile`) are usually shared among CMP transactions.

DDvO commented:
This time there is an unrelated CI failure with `90-test_threads.t`.

t8m commented:
Hmm, the added function looks like a feature. Should the PR be split?

DDvO commented:
Well, though a new function looks like a new feature, this is actually an API bug fix:
the function is needed, otherwise `OSSL_CMP_CTX_push0_geninfo_ITAV()` cannot be used properly.
ITAVs can be added, but there is no chance to remove them, so further CMP requests are e.g., stuck to use the same cert profile. Adding further `OSSL_CMP_CTX_push0_geninfo_ITAV()` calls will incorrectly add further profile names, which a server will not be able to handle. 

An alternative fix would be to tweak `OSSL_CMP_CTX_reinit()` in a similar way as done here for `genm_ITAVs`, but as I tried to explain above, this would not be appropriate for the `generalInfo` case.

t8m commented:
Then the documentation is incorrect as for the version this was added in. And the libcrypto.num update as well.

DDvO commented:
Good point - fixed.

t8m commented on util/libcrypto.num:
```diff
@@ -5425,6 +5425,7 @@ ASN1_item_d2i_ex                        5552	3_0_0	EXIST::FUNCTION:
 ASN1_TIME_print_ex                      5553	3_0_0	EXIST::FUNCTION:
 EVP_PKEY_get0_provider                  5554	3_0_0	EXIST::FUNCTION:
 EVP_PKEY_CTX_get0_provider              5555	3_0_0	EXIST::FUNCTION:
+OSSL_CMP_CTX_reset_geninfo_ITAVs        ?	3_0_0	EXIST::FUNCTION:CMP
```
This should be 3_0_6 or something.

t8m commented on doc/man3/OSSL_CMP_CTX_new.pod:
```diff
@@ -788,8 +788,7 @@ OSSL_CMP_CTX_set0_trustedStore() was renamed to OSSL_CMP_CTX_set0_trusted(),
 
-OSSL_CMP_CTX_get0_validatedSrvCert() and OSSL_CMP_CTX_reset_geninfo_ITAVs()
-were added in OpenSSL 3.1.
+OSSL_CMP_CTX_get0_validatedSrvCert() was added in OpenSSL 3.1.
```
But there needs to be an entry for the added function saying it was added in OpenSSL 3.0.6

DDvO replied:
Done as well.

DDvO commented:
Asking @openssl/committers for 2nd review.

slontis commented on doc/man3/OSSL_CMP_CTX_new.pod:
```diff
@@ -185,6 +187,7 @@ OSSL_CMP_CTX_reinit() prepares the given I<ctx> for a further transaction by
 clearing the internal CMP transaction (aka session) status, PKIStatusInfo,
 and any previous results (newCert, newChain, caPubs, and extraCertsIn)
 from the last executed transaction.
+It also clears any ITAVs added by calling OSSL_CMP_CTX_push0_genm_ITAV().
```
This sentence completely confused me, since it can be read that the reinit gets rid of ITAVS By doing a call to OSSL_CMP_CTX_push0_genm_ITAV().

Maybe..
'It also clears any ITAVs that were added by OSSL_CMP_CTX_push0_genm_ITAV().'

DDvO replied:
Re-phrased as suggested, also the other occurrence below.
@t8m, could you please re-approve.

slontis commented on doc/man3/OSSL_CMP_CTX_new.pod:
```diff
@@ -520,6 +523,9 @@ OSSL_CMP_CTX_push0_geninfo_ITAV() adds I<itav> to the stack in the I<ctx> to be
 message sent with this context.
 
+OSSL_CMP_CTX_reset_geninfo_ITAVs()
+clears any ITAVs added by calling OSSL_CMP_CTX_push0_geninfo_ITAV().
```
same here..

slontis commented on doc/man3/OSSL_CMP_CTX_new.pod:
```diff
@@ -780,6 +788,8 @@ OSSL_CMP_CTX_set0_trustedStore() was renamed to OSSL_CMP_CTX_set0_trusted(),
 using macros, while keeping the old names for backward compatibility,
 in OpenSSL 3.1.
 
+OSSL_CMP_CTX_reset_geninfo_ITAVs() was added in OpenSSL 3.0.6.
```
Not sure about the version 3.0.6 listed here is really right for master..

t8m replied:
Why not? IMO it is right. Unless we were in situation where master would be released earlier than the 3.0.6 version in which case it would be debatable. But that is definitely not the case.

DDvO replied:
I also see this as correct.
In contrast, it certainly would be wrong to state "was added in OpenSSL 3.1"  because in fact it is added already for 3.0.6.

DDvO commented:
Rebased to fix version 3.1->3.2 merge conflits.

DDvO commented:
Ping @openssl/committers for 2nd review

wan7teen-ping commented:
First step

Hi I will help you with everything you need but you have to subscribe to
the youtube channel and I will fulfill all your requests

https://youtu.be/aIaqr4eq18g

Pada Isn, 19 Sep 2022, 7:19 PM David von Oheimb ***@***.***>
menulis:

> It turns out that a similar issue was with ctx->geninfo_ITAVs,
> where in this case the solution is to add a missing API function
> OSSL_CMP_CTX_reset_geninfo_ITAVs() because the generalInfo header entries
> (such as certProfile) are usually shared among CMP transactions.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openssl/openssl/pull/19216#issuecomment-1250886812>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APESUCRQRBZDFC6HMJ65JITV7BD53ANCNFSM6AAAAAAQMRLCXM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

paulidale commented on doc/man3/OSSL_CMP_CTX_new.pod:
```diff
@@ -700,7 +706,9 @@ OSSL_CMP_certConf_cb() returns I<fail_info> if it is not equal to 0,
 or else a bit field with the B<OSSL_CMP_PKIFAILUREINFO_incorrectData> bit set.
 
-All other functions return 1 on success, 0 on error.
+All other functions, including OSSL_CMP_CTX_reinit()
```
I'm not sure the _including..._ phrase needs to be here.  It's kind of implicit in the "all other" part.

DDvO replied:
Right, the "including ..." part is implied by the rest, but I thought it could be helpful to explicitly mention this new function here, for instance for the case that a reader greps on the man page to quickly find the result behavior of that function.
Thanks @paulidale for giving the approval regardless of this point.

DDvO commented:
I'm going to resolve the trivial new merge conflict with `libcrypto.num` when merging this PR.

t8m commented:
This needs a rebase and update since 3.0.6 was already released without this.

DDvO commented:
Argh, I had missed that this was ready to merge for quite a while.
Rebased to fix the trivial conflict in `libcrypto.num` and updated the version reference.
Still ready to merge?

DDvO commented:
Looks like re-approvals are needed here?

t8m commented on util/libcrypto.num:
```diff
@@ -5430,6 +5430,7 @@ OPENSSL_strncasecmp                     5557	3_0_3	EXIST::FUNCTION:
 EVP_RAND_CTX_up_ref                     ?	3_1_0	EXIST::FUNCTION:
 RAND_set0_public                        ?	3_1_0	EXIST::FUNCTION:
 RAND_set0_private                       ?	3_1_0	EXIST::FUNCTION:
+OSSL_CMP_CTX_reset_geninfo_ITAVs        ?	3_0_7	EXIST::FUNCTION:CMP
```
3_0_8

DDvO replied:
Done

t8m commented on doc/man3/OSSL_CMP_CTX_new.pod:
```diff
@@ -780,6 +788,8 @@ OSSL_CMP_CTX_set0_trustedStore() was renamed to OSSL_CMP_CTX_set0_trusted(),
 using macros, while keeping the old names for backward compatibility,
 in OpenSSL 3.2.
 
+OSSL_CMP_CTX_reset_geninfo_ITAVs() was added in OpenSSL 3.0.7.
```
3.0.8

DDvO replied:
Done

t8m commented:
@paulidale please reconfirm

DDvO commented:
Rebased again to fix a (trivial) merge conflict in  `libcrypto.num`.

@paulidale could you please re-confirm?

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but this PR has failing CI tests.  Once the tests pass it will get moved to 'approval: ready to merge' automatically, alternatively please review and set the label manually.

DDvO commented:
Merged to master, 3.0., and 3.1.
Thanks @t8m and  @paulidale, also for the re-approvals that meanwhile were needed.

DDvO commented:
When cherry-picking to 3.0 and 3.1, I had to fix some merge conflicts regarding the new function `OSSL_CMP_CTX_reset_geninfo_ITAVs()` due to the fact that those branches do not contain functions that meanwhile were added for 3.2.
Doing so, I made a mistake deleting one more line from `OSSL_CMP_CTX_new.pod`, which made `doc-nits` fail.
I've corrected this right away in a follow-up commit.

t8m commented:
> When cherry-picking to 3.0 and 3.1, I had to fix some merge conflicts regarding the new function `OSSL_CMP_CTX_reset_geninfo_ITAVs()` due to the fact that those branches do not contain functions that meanwhile were added for 3.2.
> Doing so, I made a mistake deleting one more line from `OSSL_CMP_CTX_new.pod`, which made `doc-nits` fail.
> I've corrected this right away in a follow-up commit.

If the cherry-pick conflict resolution is not trivial a new PR should have been submitted.

DDvO commented:
> If the cherry-pick conflict resolution is not trivial a new PR should have been submitted.

Thanks for pointing this out. Will do so in the future.

In this case it was borderline between trivial and slightly more involved,
and I just made a silly mistake when deleting the automatically added line containing the `>>>>>>>`.
So it was actually independent of the complexity of the conflict resolution.
